An embedding is a dense numerical vector that captures the meaning of a word or sentence.

In [1]:
from sentence_transformers import SentenceTransformer

In [2]:
model = SentenceTransformer("all-MiniLM-L6-v2")
print("Loading model...")

embedding = model.encode("I learn Machine Learning")
print("Model loaded!")

print(type(embedding))
print(embedding.shape)

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.5k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

Loading model...
Model loaded!
<class 'numpy.ndarray'>
(384,)


**This means the sentence became a vector containing 384 numbers.**

#### Exercise 2 compare similarity

In [3]:
from sklearn.metrics.pairwise import cosine_similarity

In [5]:
s1 = model.encode('I learn AI')
s2 = model.encode('I enjoiy Artificial Intelligence')
s3 = model.encode('I hat traffic jams')

print('s1 vs s2:', 
     cosine_similarity([s1],[s2])[0][0])
print('s1 vs s3:', 
     cosine_similarity([s1],[s3])[0][0])

s1 vs s2: 0.6190995
s1 vs s3: 0.18646184


### Exercise 3 

In [7]:
sentences = ["I love coding",
             "Conding is my passion",
             "The weather is normal today",
             "Python is amazing language"
            ]
# generate embedding for all sentences at once
embeddings = model.encode(sentences)

print('embedding shape: ', embeddings.shape)

# compare every sentence with every other sentence
print("\nPairwise similarity matrix:")
similarity_matrix = cosine_similarity(embeddings)

for i in range(len(sentences)):   #use loop for all sentences
    for j in range(len(sentences)):
        if i < j :            # avoid repeating pairs and self-comparison
            print(f"'{sentences[i]}' vs '{sentences[j]}': {similarity_matrix[i][j]:.4f}")

embedding shape:  (4, 384)

Pairwise similarity matrix:
'I love coding' vs 'Conding is my passion': 0.3447
'I love coding' vs 'The weather is normal today': -0.0279
'I love coding' vs 'Python is amazing language': 0.4336
'Conding is my passion' vs 'The weather is normal today': 0.0496
'Conding is my passion' vs 'Python is amazing language': 0.2080
'The weather is normal today' vs 'Python is amazing language': -0.0061


### Exercise 4

In [9]:
import numpy as np

documents = [
    "I love programming",
    "Coding is my passion",
    "The weather is hot today",
    "Python is an amazing language",
    "I enjoy hiking in the mountains",
    "Machine learning models need data"
]

doc_embeddings = model.encode(documents)

# A search query
query = "I really like writing code"
query_embedding = model.encode(query)

# Compare query to all documents
similarities = cosine_similarity([query_embedding], doc_embeddings)[0]

# Find the best match
best_idx = np.argmax(similarities)

print(f"Query: {query}\n")
for doc, score in zip(documents, similarities):
    print(f"{score:.4f}  →  {doc}")

print(f"\nBest match: '{documents[best_idx]}' (score: {similarities[best_idx]:.4f})")

Query: I really like writing code

0.6394  →  I love programming
0.6801  →  Coding is my passion
0.0502  →  The weather is hot today
0.4652  →  Python is an amazing language
0.2177  →  I enjoy hiking in the mountains
0.1041  →  Machine learning models need data

Best match: 'Coding is my passion' (score: 0.6801)


In [10]:
documents = [
    "I love programming",
    "Coding is my passion",
    "The weather is hot today",
    "Python is an amazing language",
    "I enjoy hiking in the mountains",
    "Machine learning models need data"
]

doc_embeddings = model.encode(documents)

query = "I really like writing code"
query_embedding = model.encode(query)

similarities = cosine_similarity([query_embedding], doc_embeddings)[0]

# --- Top 3 matches ---
top_n = 3
top_indices = np.argsort(similarities)[::-1][:top_n]

print(f"Query: {query}\n")
print(f"Top {top_n} matches:")
for rank, idx in enumerate(top_indices, start=1):
    print(f"{rank}. {documents[idx]}  (score: {similarities[idx]:.4f})")

Query: I really like writing code

Top 3 matches:
1. Coding is my passion  (score: 0.6801)
2. I love programming  (score: 0.6394)
3. Python is an amazing language  (score: 0.4652)
